[Federation] Print out status updates while `kubefed init` is running

perotinus opened the pull request:
This is not an ideal final state–it does not address the appearance of hanging during long-running commands, for example–but it provides some level of information when the operations are successful.

See #41725.

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```release-note
Prints out status updates when running `kubefed init`
```

perotinus committed:
[Federation] Print out status updates while kubefed is running.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41849)
<!-- Reviewable:end -->

madhusudancs commented:
General comment for the overall change. I was thinking about doing something less verbose than this since these are user facing messages. Example:

"Creating a namespace for federation system components..."

you wait until the operation is done and then on the same line say "Ok" or "Done".

"Creating a namespace for federation system components...Done"

Something like that.

Also, don't you need linebreaks at the end of each `fmt.Fprintf()` string?

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41849#-:-KdbTz5ECcoV4AvPvRIz:b-g2bkzz)*
<!-- Sent from Reviewable.io -->

irfanurrehman commented:
*[a discussion](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41849#-KdkIUqsDxES-AyPeQrH:-KdkIUqsDxES-AyPeQrI:bldthd3) (no related file):*
I dont think the user really needs to know all the steps explicitly under the hood. The problem was that there was no indication of progress, at all, whatsoever.
My suggestion (since you have already added these prints) would be to just change the verbosity level of minor steps (might be useful for debugging) , to skip the minor steps at default  verbosity. So by default following would be printed:
1 - created namespace . . Done
2 - created service . . Done
3 - deployed api server . . Done
4 - deployed controller manager . .Done
The above are not time consuming and would happen almost immediately (seconds).

The last message could be
Waiting for the control plane to come up . . . . . . . . . . Done!
 The ". . . " as Madhu suggested could be timed prints (for example 2 seconds)

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41849)*
<!-- Sent from Reviewable.io -->

perotinus commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41849#-KdkIUqsDxES-AyPeQrH:-KdlZwRVPRjiuyprmQKF:b-3yff9i) (no related file):*
<details><summary><i>Previously, irfanurrehman wrote…</i></summary><blockquote>

I dont think the user really needs to know all the steps explicitly under the hood. The problem was that there was no indication of progress, at all, whatsoever.
My suggestion (since you have already added these prints) would be to just change the verbosity level of minor steps (might be useful for debugging) , to skip the minor steps at default  verbosity. So by default following would be printed:
1 - created namespace . . Done
2 - created service . . Done
3 - deployed api server . . Done
4 - deployed controller manager . .Done
The above are not time consuming and would happen almost immediately (seconds).

The last message could be
Waiting for the control plane to come up . . . . . . . . . . Done!
 The ". . . " as Madhu suggested could be timed prints (for example 2 seconds)


</blockquote></details>

Thanks for the feedback! Yes, that makes a lot of sense. I've created some simpler messages that will always print (and fixed the line breaks), and moved some of the more-detailed messages to verbose logs. WDYT?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41849)*
<!-- Sent from Reviewable.io -->

perotinus commented:
Ok, I've made these updates. Thanks!

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41849#-:-KdmJxerirpsBesb6_VY:b-u1gp4q)*
<!-- Sent from Reviewable.io -->

perotinus commented:
@madhusudancs @irfanurrehman PTAL?

madhusudancs commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[a discussion](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41849#-KdkIUqsDxES-AyPeQrH:-Kdr1uiaMEOpNMQnGvGb:brcyf1t) (no related file):*
<details><summary><i>Previously, perotinus (Jonathan MacMillan) wrote…</i></summary><blockquote>

Thanks for the feedback! Yes, that makes a lot of sense. I've created some simpler messages that will always print (and fixed the line breaks), and moved some of the more-detailed messages to verbose logs. WDYT?
</blockquote></details>

I have left a few comments. None of those are critical or urgent. So I would recommend making all those changes in a separate follow up PR.

In the cases where you print the messages to STDOUT, you skip logging them. I think it is worth adding an INFO log for those `creating ...` steps too. I understand why you are doing this, but if you don't have these INFO logs, people who grep for "INFO" don't see the messages printed to STDOUT.

It looks good to me otherwise. However, I would like @irfanurrehman to make another pass and add the LGTM label when he is satisfied.

---

*[federation/pkg/kubefed/init/init.go, line 241 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41849#-Kdqxuh_D49hAwXEMEHE:-Kdqxuh_D49hAwXEMEHF:b-7lpf6d) ([raw file](https://github.com/kubernetes/kubernetes/blob/3bd9ce38222d58c12558a4381752ae3c13ee87ff/federation/pkg/kubefed/init/init.go#L241)):*
> ```Go
> 
> 	// 1.
> 	fmt.Fprint(cmdOut, "Creating a namespace for federation system components... ")
> ```

Print the namespace in the message `Creating namespace "federation-system" for federation system components...`

---

*[federation/pkg/kubefed/init/init.go, line 246 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41849#-Kdr0V5vPZSwJLnJDauP:-Kdr0V5vPZSwJLnJDauQ:b6a6b9x) ([raw file](https://github.com/kubernetes/kubernetes/blob/3bd9ce38222d58c12558a4381752ae3c13ee87ff/federation/pkg/kubefed/init/init.go#L246)):*
> ```Go
> 		return err
> 	}
> 	fmt.Fprint(cmdOut, "done\n")
> ```

Use `fmt.Fprintln()`(https://golang.org/pkg/fmt/#Fprintln) instead. It's more readable. Also change all other occurrences. 

---

*[federation/pkg/kubefed/init/init.go, line 249 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41849#-Kdqy_3IMlxVw7gwnasX:-Kdqy_3IMlxVw7gwnasY:b-7d6tdi) ([raw file](https://github.com/kubernetes/kubernetes/blob/3bd9ce38222d58c12558a4381752ae3c13ee87ff/federation/pkg/kubefed/init/init.go#L249)):*
> ```Go
> 
> 	// 2.
> 	fmt.Fprint(cmdOut, "Creating federation control plane objects (service, credentials, persistent volume claim)... ")
> ```

I like this simplification a lot. However, I think service deserves its own message. It is one of the slowest operations in the entire init command, probably as slow as or slower than waiting for the control plane pods to come up.

---

*[federation/pkg/kubefed/init/init.go, line 270 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41849#-Kdqz4Pc0e0ZpsNnM2QE:-Kdqz4Pc0e0ZpsNnM2QF:b-9qox9a) ([raw file](https://github.com/kubernetes/kubernetes/blob/3bd9ce38222d58c12558a4381752ae3c13ee87ff/federation/pkg/kubefed/init/init.go#L270)):*
> ```Go
> 
> 	// 4.
> 	glog.V(4).Info("Creating an entry in the kubeconfig file with the certificate and credential data")
> ```

What's happening here is creation of kubeconfig data to be populated in a secret. While your message is correct if one reads it very carefully, it might also cause some confusion. I would just write this as `Creating credentials secret for federation controller manager`

---

*[federation/pkg/kubefed/init/init.go, line 275 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41849#-Kdqzu3nP1f4m4woyAle:-Kdqzu3nP1f4m4woyAlf:b2nreh6) ([raw file](https://github.com/kubernetes/kubernetes/blob/3bd9ce38222d58c12558a4381752ae3c13ee87ff/federation/pkg/kubefed/init/init.go#L275)):*
> ```Go
> 		return err
> 	}
> 	glog.V(4).Info("kubeconfig successfully updated")
> ```

Same as above.

---

*[federation/pkg/kubefed/init/init.go, line 298 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41849#-Kdr-7RAJnmGPw75lyWJ:-Kdr-7RAJnmGPw75lyWK:b-1k1it9) ([raw file](https://github.com/kubernetes/kubernetes/blob/3bd9ce38222d58c12558a4381752ae3c13ee87ff/federation/pkg/kubefed/init/init.go#L298)):*
> ```Go
> 
> 	// 6.
> 	fmt.Fprint(cmdOut, "Creating federation component deployments...")
> ```

I would just say `Creating federation control plane components...`

---

*[federation/pkg/kubefed/init/init.go, line 332 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41849#-Kdr1Wr4-ANMuazHDr-l:-Kdr1Wr5z8tWld3YSt2_:b-2zq3k7) ([raw file](https://github.com/kubernetes/kubernetes/blob/3bd9ce38222d58c12558a4381752ae3c13ee87ff/federation/pkg/kubefed/init/init.go#L332)):*
> ```Go
> 
> 	// 8.
> 	fmt.Fprint(cmdOut, "Updating kubeconfig... ")
> ```

No log here?

---

*[federation/pkg/kubefed/init/init.go, line 349 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41849#-Kdr1ZCD-4IrAak8G_xG:-Kdr1ZCD-4IrAak8G_xH:bql1ypa) ([raw file](https://github.com/kubernetes/kubernetes/blob/3bd9ce38222d58c12558a4381752ae3c13ee87ff/federation/pkg/kubefed/init/init.go#L349)):*
> ```Go
> 	if err != nil {
> 		return err
> 	}
> ```

Log here?

---

*[federation/pkg/kubefed/init/init.go, line 353 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41849#-Kdr1_aiFJnLlte_ZXWC:-Kdr1_aj8hXDx05llqRR:b-2zq3k7) ([raw file](https://github.com/kubernetes/kubernetes/blob/3bd9ce38222d58c12558a4381752ae3c13ee87ff/federation/pkg/kubefed/init/init.go#L353)):*
> ```Go
> 
> 	if !i.options.dryRun {
> 		fmt.Fprint(cmdOut, "Waiting for federation control plane to come up... ")
> ```

No log here?

---

*[federation/pkg/kubefed/init/init.go, line 360 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41849#-Kdr1apIwyDFiCkZUIHK:-Kdr1apIwyDFiCkZUIHL:b-6k20vm) ([raw file](https://github.com/kubernetes/kubernetes/blob/3bd9ce38222d58c12558a4381752ae3c13ee87ff/federation/pkg/kubefed/init/init.go#L360)):*
> ```Go
> 		}
> 		err = waitSrvHealthy(config, i.commonOptions.Name, i.commonOptions.Kubeconfig)
> 		fmt.Fprint(cmdOut, "done\n")
> ```

log here?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41849)*
<!-- Sent from Reviewable.io -->

irfanurrehman commented:
/lgtm

Except one thing (add it as part of follow up PR) I think everything is covered. (provided you would take the additional comments from  @madhusudancs in a follow up PR).
About printing  ". .  ", I mentioned them to be timed prints (for example with a wait of 2 seconds each), rather then they being part of the original print.
The intention is, if the step is taking long, it will sort of give a notion of a console status bar moving .. :)
You can have a common function (or a goroutine with some channel to flag when to start and when to stop) and call it at all steps.

irfanurrehman commented:
@perotinus, please run hack/update-bazel.sh to update the bazel build files and update the commit, I will lgtm it again.

madhusudancs commented:
/approve

---

Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41849#-:-KdtklAMRcUIRU8Qx_8D:bs5zpdv)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
/lgtm

perotinus commented:
@madhusudancs I ran hack/update-bazel.sh and need approval again. Can you oblige?

Thanks!

perotinus commented:
@madhusudancs Can you poke the robot again? The failures look like they were related to the BUILD file issue.

madhusudancs commented:
/lgtm

madhusudancs commented:
@k8s-bot ok to test

madhusudancs commented:
@k8s-bot gci gce e2e test this

madhusudancs commented:
/lgtm

---

Review status: 0 of 2 files reviewed at latest revision, 11 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41849#-:-KeAp8iup1F7r0JDcdzC:b-8ks5qa)*
<!-- Sent from Reviewable.io -->

perotinus commented:
@madhusudancs Can you re-LGTM this?

irfanurrehman commented:
@k8s-bot gce etcd3 e2e test this

irfanurrehman commented:
Review status: 0 of 3 files reviewed at latest revision, 12 unresolved discussions.

---

*[federation/pkg/kubefed/init/init.go, line 280 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41849#-KhD4pqH8gCxKUuvDxaF:-KhD4pqH8gCxKUuvDxaG:b947jrb) ([raw file](https://github.com/kubernetes/kubernetes/blob/8da14d4c719adbd8d3a04547ec9b7a7e17876215/federation/pkg/kubefed/init/init.go#L280)):*
> ```Go
> 
> 	// 1.
> 	fmt.Fprint(cmdOut, "Creating a namespace for federation system components... ")
> ```

You wouldn't need to print the ". . ." as part of a particular message.
Also please check my comments on https://github.com/kubernetes/kubernetes/pull/42296

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41849)*
<!-- Sent from Reviewable.io -->

perotinus commented:
Review status: 0 of 3 files reviewed at latest revision, 12 unresolved discussions.

---

*[federation/pkg/kubefed/init/init.go, line 280 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41849#-KhD4pqH8gCxKUuvDxaF:-KhPHKCMPvpwI1uzq_sR:b-jmwye3) ([raw file](https://github.com/kubernetes/kubernetes/blob/8da14d4c719adbd8d3a04547ec9b7a7e17876215/federation/pkg/kubefed/init/init.go#L280)):*
<details><summary><i>Previously, irfanurrehman wrote…</i></summary><blockquote>

You wouldn't need to print the ". . ." as part of a particular message.
Also please check my comments on https://github.com/kubernetes/kubernetes/pull/42296
</blockquote></details>

Thanks! Just replied to your comments on #42296. I'll update this PR based on the results of that discussion.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41849)*
<!-- Sent from Reviewable.io -->

perotinus commented:
Review status: 0 of 3 files reviewed at latest revision, 12 unresolved discussions.

---

*[federation/pkg/kubefed/init/init.go, line 280 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41849#-KhD4pqH8gCxKUuvDxaF:-KhhqreCoPMeVBZcvo_N:b-q09rz6) ([raw file](https://github.com/kubernetes/kubernetes/blob/8da14d4c719adbd8d3a04547ec9b7a7e17876215/federation/pkg/kubefed/init/init.go#L280)):*
<details><summary><i>Previously, perotinus (Jonathan MacMillan) wrote…</i></summary><blockquote>

Thanks! Just replied to your comments on #42296. I'll update this PR based on the results of that discussion.
</blockquote></details>

OK, based on discussion in #42296, I've updated this PR.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41849)*
<!-- Sent from Reviewable.io -->

perotinus commented:
@madhusudancs @irfanurrehman I think this should be good-to-go at this point. Can someone LGTM?

irfanurrehman commented:
A very minor nit, else this pr is good to go from me.
Also, apologies for being this picky, I guess user facing code tends to attract lot of different opinions from reviewers... !

---

Review status: 0 of 3 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[federation/pkg/kubefed/init/init.go, line 280 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41849#-KhD4pqH8gCxKUuvDxaF:-KhugSVTUXPizzpy5IEa:bf91tyy) ([raw file](https://github.com/kubernetes/kubernetes/blob/8da14d4c719adbd8d3a04547ec9b7a7e17876215/federation/pkg/kubefed/init/init.go#L280)):*
<details><summary><i>Previously, perotinus (Jonathan MacMillan) wrote…</i></summary><blockquote>

OK, based on discussion in #42296, I've updated this PR.
</blockquote></details>

Thanks @perotinus. I see that you have removed all the dots from the messages.
If the operation finishes fast (before the timed print of ```. .``` happens ```done``` will be printed right next to the last message, without even space.. :-)
For example ```Creating a namespace for federation system componentsdone```

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41849#-:-Khuhs1M8mv7PszSaPr9:b33ihri)*
<!-- Sent from Reviewable.io -->

perotinus commented:
No worries! Better to have it be right, especially when it'll be exposed to users.

---

Review status: 0 of 3 files reviewed at latest revision, 12 unresolved discussions.

---

*[federation/pkg/kubefed/init/init.go, line 280 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41849#-KhD4pqH8gCxKUuvDxaF:-KhxuTRtJ0kddfjUYaOS:b6e5uee) ([raw file](https://github.com/kubernetes/kubernetes/blob/8da14d4c719adbd8d3a04547ec9b7a7e17876215/federation/pkg/kubefed/init/init.go#L280)):*
<details><summary><i>Previously, irfanurrehman wrote…</i></summary><blockquote>

Thanks @perotinus. I see that you have removed all the dots from the messages.
If the operation finishes fast (before the timed print of ```. .``` happens ```done``` will be printed right next to the last message, without even space.. :-)
For example ```Creating a namespace for federation system componentsdone```
</blockquote></details>

Ah, yes, that make sense. Fixed.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41849#-:-KhxuLoXmTRR0vs9ptoz:b-ohusce)*
<!-- Sent from Reviewable.io -->

irfanurrehman commented:
/lgtm
